### PR TITLE
fix(core): ensure workspace version format is respected when making changes to it

### PR DESCRIPTION
**Pull request by leosvelperez**
## Current Behavior

When working with generators and the Nx DevKit utilities to make changes to a workspace configuration that it's using the v1 format, the format of the workspace configuration is changed to v2.

A workaround is to call `await formatFiles(tree);` at the end of the generation because this utility makes sure to leave the workspace configuration in the right format. That's not an intuitive solution. We already have in place utilities that make sure that when we read the configuration it will be returned in a consistent way (v2 format), but we don't do the same when we write the configuration to make sure is restored to the original version.

## Expected Behavior

Updating the Nx configuration files should be straightforward and shouldn't require us to do any additional steps to make sure the configuration is written in the right format. The interoperability between the different workspace versions should be transparent to the end-user.

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/linter/src/utils/convert-tslint-to-eslint/__snapshots__/project-converter.spec.ts.snap [hunk 1]
```diff
@@ -94,6 +72,28 @@ Object {
...
+    "@nrwl/angular": Object {
+      "application": Object {
+        "linter": "tslint",
+      },
+      "library": Object {
+        "linter": "tslint",
+      },
     },
   },
   "version": 1,
```
Comment (AgentEnder) [inline packages/linter/src/utils/convert-tslint-to-eslint/__snapshots__/project-converter.spec.ts.snap:99]: Wow, its weird seeing a snapshot that demonstrates the exact issue you were fixing. We should have caught this sooner 😦

**Review by AgentEnder — APPROVED** [review]
Looks good, thanks for getting this fixed Leo.

**Comment by AgentEnder** [comment]
I'll wait to give @FrozenPandaz a chance to review, he may catch something I missed

File: packages/linter/src/utils/convert-tslint-to-eslint/__snapshots__/project-converter.spec.ts.snap (hunk 1 above)
Comment (leosvelperez) [reply packages/linter/src/utils/convert-tslint-to-eslint/__snapshots__/project-converter.spec.ts.snap:99]: Yeah, the thing is all our tests use the `createTreeWithEmptyWorkspace` with no parameter and that creates the workspace with version 1. We might need to be mindful moving forward when we write tests to make sure it works well across the different versions.

Comment (leosvelperez) [reply packages/linter/src/utils/convert-tslint-to-eslint/__snapshots__/project-converter.spec.ts.snap:99]: But I agree, this particular test case was overlooked. It was starting as a v1 workspace but it ended up as a v2 workspace and we didn't notice before.

File: packages/workspace/src/generators/init/init.spec.ts [hunk 2]
```diff
@@ -56,6 +56,10 @@ describe('workspace', () => {
         '/tsconfig.spec.json',
         '{"extends": "../tsconfig.json", "compilerOptions": {}}'
       );
+      tree.write(
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/init/init.spec.ts:59]: This looks completely separate? Can updates to `init.ts` be in a separate PR?

File: packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts [hunk 3]
```diff
@@ -190,6 +196,8 @@ describe('moveProjectConfiguration', () => {
   });
 
   it('should support moving a standalone project', () => {
+    setupWorkspace(2);
```
Comment (FrozenPandaz) [inline packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts:199]: This was because the function now throws when `standalone` is `true` but the workspace version is 1?

File: packages/devkit/src/generators/project-configuration.spec.ts [hunk 4]
```diff
@@ -24,353 +40,519 @@ const baseTestProjectConfig: ProjectConfiguration = {
...
       });
-      updateJson(tree, 'nx.json', (json) => {
-        json.projects['proj1'] = {
-          tags: ['tag1'],
-        };
-        return json;
+    });
+
+    describe('addProjectConfiguration', () => {
+      it('should throw when standalone is true', () => {
```
Comment (FrozenPandaz) [inline packages/devkit/src/generators/project-configuration.spec.ts:104]: It seems like this didn't throw before... why does it throw now?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/workspace/src/generators/init/init.spec.ts (hunk 2 above)
Comment (leosvelperez) [reply packages/workspace/src/generators/init/init.spec.ts:59]: It's not separate at all. The `init` generator is broken after the change introduced in this PR. It's broken for a couple of reasons:

- This generator uses `readJson` to read the configuration files instead of the `readProjectConfiguration` or similar utilities that normalize the result to the v2 format.
- Because of the above and because the update configuration methods were always writing v2 format, there were pieces of code using `architect` and others using `targets` based on when the update happened.

So, given that updating the configuration will now always write in the right format. I needed to either keep using the `readJson` and change everything to use `architect` or start using the devkit utilities and use `targets`. I did the latter to ensure consistency.

File: packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts (hunk 3 above)
Comment (leosvelperez) [reply packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts:199]: Exactly. The test tree is created by default using v1, but right after the creation in the `beforeEach` we call `addProjectConfiguration` which was turning it into v2, but with the change on this PR, it will continue to be v1, so this test case fails if we don't specify the workspace to use v2.

File: packages/devkit/src/generators/project-configuration.spec.ts (hunk 4 above)
Comment (leosvelperez) [reply packages/devkit/src/generators/project-configuration.spec.ts:104]: Before, nothing in `addProjectConfiguration` was making the check to disallow standalone projects in v1 workspaces. That check was happening when calling `formatFiles` which was the one calling `reformattedWorkspaceJsonOrNull` which calls `toOldFormatOrNull` when the workspace is v1. That last function is the one making the check.

With the changes in this PR, as part of any function to add/update the configuration, the `reformattedWorkspaceJsonOrNull` is called so the configuration is written in the right format and therefore the check happens.

**Comment by leosvelperez** [comment]
@FrozenPandaz do you have any remaining concerns about this PR? Please, let me know if there's anything else to do, I replied to your questions in the comments. It'd be good to get this in.

File: packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts [hunk 5]
```diff
@@ -24,7 +24,7 @@ describe('moveProjectConfiguration', () => {
       relativeToRootDestination: 'apps/subfolder/my-destination',
     };
 
-    tree = createTreeWithEmptyWorkspace();
+    tree = createTreeWithEmptyWorkspace(version);
```
Comment (edbzn) [inline packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts:27]: Remark related to this, would it be preferable to set the default version of `createTreeWithEmptyWorkspace(version = 1)` to `2` instead of `1` since v2 format is the most used in the repo?

**Review by edbzn — APPROVED** [review]

File: packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts (hunk 5 above)
Comment (leosvelperez) [reply packages/workspace/src/generators/move/lib/move-project-configuration.spec.ts:27]: Yeah, we need to make that change but not as part of this PR. It's a breaking change since this utility is exposed as part of the devkit and folks could be relying on the v1 format. We'll discuss this and make the change in a separate PR ensuring it's included in the release notes as a breaking change.

**Review by FrozenPandaz — APPROVED** [review]
